pkg/github/token.go

```diff
@@ -0,0 +1,91 @@
+//
+// Copyright 2023 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package github
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gofri/go-github-ratelimit/github_ratelimit"
+	"github.com/google/go-github/v41/github"
+	"golang.org/x/oauth2"
+)
+
+type GitHubToken interface {
+	GetNewGitHubClient() (*github.Client, error)
+}
+
+type GitHubTokenClient struct {
+}
+
+var Tokens []string
+
+// ParseGitHubTokens parses all of the possible GitHub tokens available to HAS and makes them available within the "github" package
+// This function should *only* be called once: at operator startup.
+func ParseGitHubTokens() error {
+	githubToken := os.Getenv("GITHUB_AUTH_TOKEN")
+	githubTokenList := os.Getenv("GITHUB_TOKEN_LIST")
```

Comment:
Is there a way we can make this a key value pair list?  It's easier to have an identifier associated with the token in order to log metrics about its usage

Reply:
Updated as discussed on today's call.

`Tokens` will be a map where the key points to the token name, and the value is the token itself. For tokens set through the legacy GITHUB_AUTH_TOKEN environment variable (which does not used named tokens), the key `GITHUB_AUTH_TOKEN` is used.

Reply:
I added it in an earlier commit here: https://github.com/redhat-appstudio/application-service/blob/7c07b8d5803ae91781b7155378a64cac36036404/README.md#creating-a-github-secret-for-has